File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/utils/Destroyer.java
package ch.rmy.android.http_shortcuts.utils;

import java.util.HashSet;
import java.util.Set;

public class Destroyer implements Destroyable {

    private final Set<Destroyable> destroyables = new HashSet<>();

    public <T extends Destroyable> T own(T destroyable) {
        destroyables.add(destroyable);
        return destroyable;
    }

    @Override
    public void destroy() {
        for (Destroyable destroyable : destroyables) {
            destroyable.destroy();
        }
        destroyables.clear();
    }

}

File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/utils/Destroyer.kt
package ch.rmy.android.http_shortcuts.utils

class Destroyer : Destroyable {

    private val destroyables = mutableListOf<Destroyable>()

    fun <T> own(destroyable: T): T where T : Destroyable {
        destroyables.add(destroyable)
        return destroyable
    }

    fun own(destroyable: () -> Unit): Destroyable {
        val destroyableObject = object : Destroyable {
            override fun destroy() {
                destroyable.invoke()
            }
        }
        destroyables.add(destroyableObject)
        return destroyableObject
    }

    override fun destroy() {
        for (destroyable in destroyables) {
            destroyable.destroy()
        }
        destroyables.clear()
    }

}
